ДЗ 12: Web сервер

anv0l opened the pull request:
Цель:
Научиться создавать серверный и пользовательский http-интерфейсы. Научиться встраивать web-сервер в уже готовое приложение.

Встроить веб-сервер в приложение из ДЗ про Hibernate ORM (или в пример из вебинара встроить ДЗ про Hibernate :)).

Сделать стартовую страницу, на которой админ должен аутентифицироваться. Сделать админскую страницу для работы с клиентами. На этой странице должны быть доступны следующие функции:

* создать клиента
* получить список клиентов

anv0l committed:
ДЗ 12: Web сервер

Цель:
Научиться создавать серверный и пользовательский http-интерфейсы. Научиться встраивать web-сервер в уже готовое приложение.

Встроить веб-сервер в приложение из ДЗ про Hibernate ORM (или в пример из вебинара встроить ДЗ про Hibernate :)).

Сделать стартовую страницу, на которой админ должен аутентифицироваться. Сделать админскую страницу для работы с клиентами. На этой странице должны быть доступны следующие функции:

* создать клиента
* получить список клиентов

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/server/ClientsWebServerHomework.java:
```diff
@@ -0,0 +1,149 @@
+    protected final TemplateProcessor templateProcessor;
+    private final Server server;
+    private final UserAuthService authService;
+    private DbServiceClientImpl dbServiceClient;
```
Это лучше получит извне, через конструктор и по интерфейсу. При текущем варианте придется менять класс сервера если изменится реализация `DbServiceClient`

anv0l replied:
Сделал инициализацию через конструктор.

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/server/ClientsWebServerHomework.java:
```diff
@@ -0,0 +1,149 @@
+    private static final String COMMON_RESOURCES_DIR = "static";
+    public static final String HIBERNATE_CFG_FILE = "hibernate.cfg.xml";
+
+    private final UserDao userDao;
```
Зачем тут нужен `UserDao`?

anv0l replied:
Уже не нужен, остался от старого кода, убрал.

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/server/ClientsWebServerHomework.java:
```diff
@@ -0,0 +1,149 @@
+        return servletContextHandler;
+    }
+
+    private void InitDB() {
+        var configuration = new Configuration().configure(HIBERNATE_CFG_FILE);
+
+        var dbUrl = configuration.getProperty("hibernate.connection.url");
+        var dbUserName = configuration.getProperty("hibernate.connection.username");
+        var dbPassword = configuration.getProperty("hibernate.connection.password");
+
+        var sessionFactory = HibernateUtils.buildSessionFactory(configuration,
+                Client.class,
+                AddressDataSet.class,
+                PhoneDataSet.class);
+
+        var transactionManager = new TransactionManagerHibernate(sessionFactory);
+        var clientTemplate = new DataTemplateHibernate<>(Client.class);
+        this.dbServiceClient = new DbServiceClientImpl(transactionManager, clientTemplate);
+    }
```
Вряд ли это ответственность веб-сервера (сборка приложения, конфигурация сервисов)

anv0l replied:
Перенёс в класс Homework. Или где DbServiceClient инициализировать?

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/server/ClientsWebServerHomework.java:
```diff
@@ -0,0 +1,149 @@
+        this.dbServiceClient = new DbServiceClientImpl(transactionManager, clientTemplate);
+    }
+
+    private void initData() {
+        for (var i = 1; i <= 2; i++) {
+            var clientAddress = new AddressDataSet();
+            var client = new Client("Client #" + i);
+            clientAddress.setStreet("#" + i  + " Client address", client);
+            var clientPhones = new ArrayList<PhoneDataSet>();
+            for (int j = 1; j <= 5; j++) {
+                clientPhones.add(new PhoneDataSet("number" + j * i, client));
+            }
+
+            client.setAddress(clientAddress);
+            client.setPhones(clientPhones);
+
+            dbServiceClient.saveClient(client);
+        }
+
+        var clients = dbServiceClient.findAll();
+        Type dataType = new TypeToken<ArrayList<Client>>(){}.getType();
+        System.out.println(gson.toJson(clients, dataType));
+    }
```
И это) Но инициализацию базы начальными данными лучше сделать через системы миграции

anv0l replied:
Убрал инициализацию начальными данными вообще.

stvort commented on hw12-webserver/webserver/src/main/resources/hibernate.cfg.xml:
```diff
@@ -0,0 +1,23 @@
+    <property name="hibernate.show_sql">true</property>
+    <property name="hibernate.generate_statistics">true</property>
+
+    <property name="hibernate.hbm2ddl.auto">update</property>
```
Как и инициализацию схемы

anv0l replied:
Схему оставил как есть, её инициализацией занимается Hibernate, осталось от 10 ДЗ.

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/servlet/ClientsServlet.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        var clients = dbServiceClient.findAll();
+        Map resultMap = new HashMap();
```
```
Map<String, Object> resultMap = new HashMap<>();
```

anv0l replied:
Поправил.

stvort commented on hw12-webserver/webserver/src/main/resources/templates/clients.html:
```diff
@@ -0,0 +1,82 @@
+            <td>Адрес:
+            <td><input id="address" name="address" type="text">
+        </tr>
+        <tr>
```
Нет парного тега

anv0l replied:
Добавил закрывающие теги. TR и TD по привычке не закрываю.

stvort commented on hw12-webserver/webserver/src/main/resources/templates/clients.html:
```diff
@@ -0,0 +1,82 @@
+            <td><input id="address" name="address" type="text">
+        </tr>
+        <tr>
+            <td>Телефоны:
```
И тут

anv0l replied:
Тут тоже добавил.

stvort commented on hw12-webserver/webserver/src/main/resources/templates/clients.html:
```diff
@@ -0,0 +1,82 @@
+        </tr>
+        <tr>
+            <td>Телефоны:
+            <td>
```
И тут)

anv0l replied:
И сюда добавил.

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/servlet/ClientsServlet.java:
```diff
@@ -0,0 +1,43 @@
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        var clients = dbServiceClient.findAll();
+        Map resultMap = new HashMap();
+        resultMap.put(ALL_CLIENTS_PARAM, gson.toJson(clients));
```
А для чего клиентов в JSON нужно преобразовывать? Можно прямо список положить

anv0l replied:
Да что-то я запутался, что может в мапу с результатами идти, думал, что должна быть именно строка. Исправил на объект - в шаблоне от ?eval избавился.

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/servlet/ClientsApiServlet.java:
```diff
@@ -0,0 +1,71 @@
+        this.gson = gson;
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.sendRedirect("/clients");
+    }
```
Зачем так нужно? Нет такого эндпойнта и нет)

anv0l replied:
Удалил этот сервлет.

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/servlet/ClientsApiServlet.java:
```diff
@@ -0,0 +1,71 @@
+        String id = (path.length > 1)? path[ID_PATH_PARAM_POSITION]: String.valueOf(- 1);
+
+        switch (id) {
+            case ("add"):{
```
По REST в апи (урлах) действия не указываются. Действие определяет HTTP-метод

anv0l replied:
Не понял, что надо сделать, поэтому удалил этот сервлет вообще.

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/servlet/ClientsApiServlet.java:
```diff
@@ -0,0 +1,71 @@
+                break; // operation not supported
+        }
+
+        response.sendRedirect("/clients");
```
Апи это про данные. Оно их принимает и/или отдает. Но не решает, что запрашивающий с ними будет делать и куда пойдет после запроса. В т.ч. поэтому апи могут пользоваться разные клиенты. От браузеров до мобилок и десктопов. Так что либо вставлять надо не через форму, а через JS, либо возможно лучше сделать обычный, не `Api` сервлет

anv0l replied:
Опять же, не понял, как должно быть, сделал добавление через POST в clients.

anv0l committed:
Исправления:

- Убрал лишние свойства - userDao и gson.
- Поле dbServiceClient в ClientsWebServerHomework теперь инициализируется через конструктор.
- Инициализацию БД перенёс в класс Homework.
- Убрал инициализацию начальными данными.
- Убрал сервлет ClientsApiServlet, теперь добавление клиента через метод POST в /clients.
- Добавил закрывающие теги для TR/TD в шаблон clients.html.
- Добавил вывод сообщения на странице с клиентами в случае, если переданы пустые значения для имени или адреса.

stvort commented on hw12-webserver/webserver/src/main/java/ru/otus/servlet/ClientsServlet.java:
```diff
@@ -0,0 +1,87 @@
+            default -> resultMap.put(PARAM_ALERT_MESSAGE, "Действие неопределено");
+        }
+
+        findAllClientsAndPutInResultMap(resultMap);
```
Тут, как раз лучше сделать редирект, а то останемся на `/add`
